brewcop.py

```diff
@@ -217,64 +205,238 @@ class Brewcop:
+
+
+class Brains:
+    """
+    Add some scale memory and semantics for interpreting a series
+    of weights as human activity.
+
+    Implement a state machine consisting of the following states:
+    unknown - no scale readings stored yet
+    brewing - scale readings have shown some increase over last 30s
+    ready - scale readings are stable/decreasing and pot still has content
+    empty - scale readings are stable/decreasing and pot content is low
+    """
+
+    """Retain scale samples for history_length seconds"""
+    history_length = 30
+
+    def __init__(self, tick_period=1, empty_thresh=0, stale_thresh=60*60*8):
+        self.history = deque(maxlen=int(self.history_length / tick_period))
+        self.pot_empty_thresh_g = empty_thresh
+        self.stale_thresh = stale_thresh
+        self.state = "unknown"
+        self.timestamp = 0
+
+    def notify(self):
+        """Stub for slack notification"""
+        return
+
+    def increasing(self, l):
+        """
+        Return true if any of l[i] > l[i + 1].
+        N.B. 'l' must be a list as deques don't allow slicing
+        N.B. Ignores l[i] < l[i + 1].
+        """
+        return any(x > y for x, y in zip(l, l[1:]))
+
+    def brewcheck(self):
+        """
+        Process new scale reading, transitioning state, if needed.
+        Call notify() on brewing->ready state transition.
+        """
+        if self.increasing(list(self.history)):
+            if self.state != "brewing":
+                self.state = "brewing"
+                self.timestamp = time.time()
+        elif self.history[0] <= self.pot_empty_thresh_g:
+            if self.state != "empty":
+                self.state = "empty"
+                self.timestamp = time.time()
+        else:
+            if self.state == "brewing": # only notify on brewing->ready
+                self.notify()
+            if self.state != "ready":
+                self.state = "ready"
+                self.timestamp = time.time()
+
+    def store(self, w):
+        """Record a scale measurement"""
+        self.history.appendleft(w)
+        self.brewcheck()
+
+    def timestr(self, t):
+        """Return a human-friendly string representing elapsed time t"""
+        daysecs = 60*60*24
+        if t < daysecs:
+            return time.strftime("%H:%M:%S", time.gmtime(t))
+        elif t < daysecs*2:
+            return "1 day"
+        else:
+            return "{} days".format(int(t/daysecs))
+
+    @property
+    def display(self):
+        """Get message text describing the state, with time since entered"""
+        t = time.time() - self.timestamp
+        timestr = self.timestr(t)
+        if self.state == "brewing":
+            return ("red", "Brewing, elapsed: {}".format(timestr))
+        elif self.state == "ready" and t < self.stale_thresh:
+            return ("green", "Ready, elapsed: {}".format(timestr))
+        elif self.state == "ready":
+            return ("red", "Ready, elapsed: {}".format(timestr))
```

Comment:
Should this say stale?  Or maybe `Ready (stale)`?

I guess the time is displayed so maybe this isn't necessary.  The color change is a nice touch!